src/Components/Components/src/BindConverter.cs

```diff
@@ -488,6 +488,182 @@ private static string FormatDateTimeOffsetValueCore(DateTimeOffset value, Cultur
             return value.Value.ToString(culture ?? CultureInfo.CurrentCulture);
         }
 
+        /// <summary>
+        /// Formats the provided <paramref name="value"/> as a <see cref="System.String"/>.
+        /// </summary>
+        /// <param name="value">The value to format.</param>
+        /// <param name="culture">
+        /// The <see cref="CultureInfo"/> to use while formatting. Defaults to <see cref="CultureInfo.CurrentCulture"/>.
+        /// </param>
+        /// <returns>The formatted value.</returns>
+        [SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Required to maintain compatibility")]
+        public static string FormatValue(DateOnly value, CultureInfo? culture = null) => FormatDateOnlyValueCore(value, format: null, culture);
+
+        /// <summary>
+        /// Formats the provided <paramref name="value"/> as a <see cref="System.String"/>.
+        /// </summary>
+        /// <param name="value">The value to format.</param>
+        /// <param name="format">The format to use. Provided to <see cref="DateOnly.ToString(string, IFormatProvider)"/>.</param>
+        /// <param name="culture">
+        /// The <see cref="CultureInfo"/> to use while formatting. Defaults to <see cref="CultureInfo.CurrentCulture"/>.
+        /// </param>
+        /// <returns>The formatted value.</returns>
+        [SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Required to maintain compatibility")]
+        public static string FormatValue(DateOnly value, string format, CultureInfo? culture = null) => FormatDateOnlyValueCore(value, format, culture);
+
+        private static string FormatDateOnlyValueCore(DateOnly value, string? format, CultureInfo? culture)
+        {
+            if (format != null)
+            {
+                // We convert to a DateTime so formatting doesn't throw if the format includes time information
+                return value.ToDateTime(TimeOnly.MinValue).ToString(format, culture ?? CultureInfo.CurrentCulture);
```

Comment:
We don't have to do this if we decide not to support binding a value of type `DateOnly` to inputs with time information (e.g. `datetime-local`). Same question applies for `TimeOnly`.

Reply:
It seems like we'd be hiding a possibly misconfigured app if we had this. What does the error experience look like if we did not do this? Does it throw a meaningful exception?

Reply:
I kind of like the flexibility that you can use any UI style with any underlying data type, and we always do the best available conversion (including discarding information when applicable). If there were cases where this would actively surprise developers or give them incorrect results then we should reconsider, but it seems like we're always giving the developer the most information we can based on the data type they've chosen.

Or am I misunderstanding? Are there any cases where people might get undesirable results?

Reply:
One thing it would be great to check we've thought through is timezones. If you bind a `type="month"` to a `DateTime`, does it show up as a local datetime or a UTC one? Does the browser supply a string that represents a specific timezone (e.g., with the `Z` suffix for UTC) or does it just ambiguously say something like `2021-08` and if so what kind of `DateTime` do we supply?

I would *hope* that the `DateTime` would represent midnight at the start of the month in UTC, and equivalent for picking a day. For `TimeOnly`, it looks like there is no timezone so that's probably correct, but again it would be good to check that even if you bind a `type="datetime-local"` to a `TimeOnly` that the hour/minute numbers entered by the user are what show up as the hour/minute numbers on the `TimeOnly` regardless of the browser and server timezones.

If there's anything difficult or risky about any of these cases, then I'd definitely be open to just not supporting them.

Reply:
The exception that gets thrown is from `DateOnly.ToString()`:

```
Unhandled exception rendering component: Input string was not in a correct format.
      System.FormatException: Input string was not in a correct format.
         at System.DateTimeFormat.IsValidCustomDateFormat(ReadOnlySpan`1 format, Boolean throwOnError) in System.Private.CoreLib.dll:token 0x600106e+0x124
         at System.DateOnly.ToString(String format, IFormatProvider provider) in System.Private.CoreLib.dll:token 0x6000dda+0xf1
         at Microsoft.AspNetCore.Components.BindConverter.FormatDateOnlyValueCore(DateOnly value, String format, CultureInfo culture) in C:\aspnetcore\src\Components\Components\src\BindConverter.cs:line 520
...
```

Not ideal perhaps but it would probably get the idea across about what the problem was. I'm good with changing this functionality if we know there isn't a scenario where you would want to bind a `DateOnly` to a `type=datetime-local` input, etc.

Reply:
There's a couple places in the framework that catch FormatException and wrap it in a more specific error. I'd be open to that.

Reply:
I didn't mean to ignore your comments, @SteveSandersonMS - I had to refresh the page for them to show up 😅 

The browser just supplies a "yyyy-MM" string for `type="month"` inputs, and it does appear that the `DateTime` we get is the start of the month at midnight. Same also applies for picking a day for `type="date"`. Since `DateTime` doesn't have time zones (I believe that's what `DateTimeOffset` would be used for), I don't think we'll have time zone issues here.

I personally don't see it as being a problem to support these cases. But, if they truly only occur when the developer has made a mistake by specifying the wrong input type or binding to the wrong date/time type, maybe it would be better to throw an error 🤷‍♂️